File: 3956-maximum-sum-of-m-non-overlapping-subarrays-i/3956-maximum-sum-of-m-non-overlapping-subarrays-i.go
func maximumSum(nums []int, m int, l int, r int) int64 {
	unboundDP := func(pre []int64, n, l, r int) int64 {
		dp := make([]int64, n+1)
		dq := make([]int, 0)
		ans := int64(-1 << 60)

		for i := l; i <= n; i++ {
			t := i - l
			cur := dp[t] - pre[t]

			for len(dq) > 0 && dp[dq[len(dq)-1]]-pre[dq[len(dq)-1]] <= cur {
				dq = dq[:len(dq)-1]
			}

			dq = append(dq, t)

			for len(dq) > 0 && dq[0] < i-r {
				dq = dq[1:]
			}

			dp[i] = dp[i-1]

			if len(dq) > 0 {
				v := pre[i] + dp[dq[0]] - pre[dq[0]]
				if v > dp[i] {
					dp[i] = v
				}
				if v > ans {
					ans = v
				}
			}
		}

		return ans
	}

	n := len(nums)

	pre := make([]int64, n+1)
	for i := 0; i < n; i++ {
		pre[i+1] = pre[i] + int64(nums[i])
	}

	neg := int64(-1 << 60)

	if m >= n/l {
		return unboundDP(pre, n, l, r)
	}

	prev := make([]int64, n+1)

	ans := neg
	last := neg

	for k := 1; k <= m; k++ {
		cur := make([]int64, n+1)
		for i := range cur {
			cur[i] = neg
		}

		dq := make([]int, 0)

		for i := l; i <= n; i++ {
			t := i - l
			val := prev[t] - pre[t]

			for len(dq) > 0 && prev[dq[len(dq)-1]]-pre[dq[len(dq)-1]] <= val {
				dq = dq[:len(dq)-1]
			}

			dq = append(dq, t)

			for len(dq) > 0 && dq[0] < i-r {
				dq = dq[1:]
			}

			if len(dq) > 0 {
				cur[i] = pre[i] + prev[dq[0]] - pre[dq[0]]
				if cur[i] > ans {
					ans = cur[i]
				}
			}
		}

		if ans == last {
			break
		}

		last = ans

		best := make([]int64, n+1)
		for i := range best {
			best[i] = neg
		}

		for i := 1; i <= n; i++ {
			if best[i-1] > cur[i] {
				best[i] = best[i-1]
			} else {
				best[i] = cur[i]
			}
		}

		prev = best
	}

	return ans
}

File: 3956-maximum-sum-of-m-non-overlapping-subarrays-i/3956-maximum-sum-of-m-non-overlapping-subarrays-i.swift
class Solution {
    func maximumSum(_ nums: [Int], _ m: Int, _ l: Int, _ r: Int) -> Int {

        func unboundDP(_ pre: [Int64], _ n: Int, _ l: Int, _ r: Int) -> Int64 {
            var dp = Array(repeating: Int64(0), count: n + 1)
            var dq = [Int]()
            var head = 0
            var ans = Int64.min / 2

            for i in l...n {
                let t = i - l
                let cur = dp[t] - pre[t]

                while dq.count > head &&
                      dp[dq.last!] - pre[dq.last!] <= cur {
                    dq.removeLast()
                }

                dq.append(t)

                while dq.count > head && dq[head] < i - r {
                    head += 1
                }

                dp[i] = dp[i - 1]

                if dq.count > head {
                    let v = pre[i] + dp[dq[head]] - pre[dq[head]]
                    dp[i] = max(dp[i], v)
                    ans = max(ans, v)
                }
            }

            return ans
        }

        let n = nums.count
        var pre = Array(repeating: Int64(0), count: n + 1)

        for i in 0..<n {
            pre[i + 1] = pre[i] + Int64(nums[i])
        }

        let neg = Int64.min / 2

        if m >= n / l {
            return Int(unboundDP(pre, n, l, r))
        }

        var prev = Array(repeating: Int64(0), count: n + 1)
        var ans = neg
        var last = neg

        for _ in 1...m {

            var cur = Array(repeating: neg, count: n + 1)
            var dq = [Int]()
            var head = 0

            for i in l...n {

                let t = i - l
                let val = prev[t] - pre[t]

                while dq.count > head &&
                      prev[dq.last!] - pre[dq.last!] <= val {
                    dq.removeLast()
                }

                dq.append(t)

                while dq.count > head && dq[head] < i - r {
                    head += 1
                }

                if dq.count > head {
                    cur[i] = pre[i] + prev[dq[head]] - pre[dq[head]]
                    ans = max(ans, cur[i])
                }
            }

            if ans == last {
                break
            }

            last = ans

            var best = Array(repeating: neg, count: n + 1)

            for i in 1...n {
                best[i] = max(best[i - 1], cur[i])
            }

            prev = best
        }

        return Int(ans)
    }
}